proposals/0000-arbitrary-expressions-in-selectors.md

```diff
@@ -0,0 +1,107 @@
+* Review manager: TBD
+
+## Introduction
+
+It is currently possible to write arbitrary expressions inside `#selector` like the following: `#selector(callThisFunc().bar)`. This complicates the implementation of proposals [SE-0064](https://github.com/apple/swift-evolution/blob/master/proposals/0064-property-selectors.md) (*Referencing Objective-C selector of property getters and setters*) and [SE-0062](https://github.com/apple/swift-evolution/blob/master/proposals/0062-objc-keypaths.md) (*Referencing Objective-C key-paths*) **a lot**. 
+
+This proposal restricts expressions inside selectors to be a sequence of property or method refernces. I believe this will not be a major restrictions since arbitrary expressions in selectors are probably rarely used, have some rough edges and removing them would simplify the compiler.
+
+## Proposed solution
+
+I propose allowed expressions inside `#selector` (and once implemented `#keyPath`) to be a series of instance or class members separated by `.` and allow disambiguating the last component using `as`.
+
+## Detailed design
+
+### Examples
+
+```swift
+class Address: NSObject {
+  dynamic var street: String
+  dynamic var town: String
+  
+  init(street: String, town: String) {
+    self.street = street
+    self.town = town
+  }
+}
+
+class Person: NSObject {
+  dynamic var name: String
+  dynamic var homeAddress: Address
+  
+  func workAddress() -> Address {
+    // ...
+  }
+  
+  func workAddress(formatter: AddressFormatter) -> String {
+    // ...
+  }
+
+  init(name: String, homeAddress: Address) {
+    self.name = name
+    self.homeAddress = homeAddress
+  }
+}
+
+let me: Person = ...
+```
+
+The following examples will continue to work:
+
+```swift
+let _ = #selector(getter: Person.name)
+let _ = #selector(getter: me.name)
+
+let _ = #selector(getter: Person.homeAddress.street)
+// Could also be written as
+let _ = #selector(getter: Address.street)
+
+let _ = #selector(Person.workAddress as () -> Address)
+let _ = #selector(Person.workAddress(formatter: ))
+```
+
+I propose removing this kind of selector:
+
+```swift
+// Should produce selector "street". Note that the method workAddress() is never 
+// called and its return type only used during type checking
+let _ = #selector(getter: me.workAddress().street) 
+
+// The above can be rewritten in a cleaner way like the following
+let _ = #selector(getter: Address.street)
+```
+
+The proposed way to rewrite the selector elimininates potential confusion about the fact that calling a method inside `#selector` actually doesn't invoke it.
+
+### Grammar
+
+<pre>
+selector → #selector(<i>selector-modifer<sub>opt</sub></i> <i>selector-path</i>)
+
+selector-modifier → getter:
+selector-modifier → setter:
+
+selector-path → <i>type-identifier</i> . <i>selector-member-path</i> <i>as-disambiguation<sub>opt</sub></i>
+selector-path → <i>selector-member-path</i> <i>as-disambiguation<sub>opt</sub></i>
+
+selector-member-path → <i>identifier</i>
```

Comment:
Can we add an optional "()" after the identifier production of selector-member-path? That would cover the zero-parameter case